---
title: "16S_Microbial_Community_Analysis"
author: "Dan Shelley"
date: "`r Sys.Date()`"

output:
  html_document:
    theme: flatly
    css: style.css
    toc: true
    toc_float: true
    toc_depth: 3
---

```{r defining_user_variables, include = FALSE}

analysis_name <- "Hurleston Reservoir, United Utilities"

## Input Files from Qiime2 Pipeline
metadata_path <- "metadata.csv"
otu_table_path <- "feature-table.tsv"
otu_fasta_path <- "dna-sequences.fasta"
taxonomy_table_path <- "taxonomy.tsv"
tree_path <- "tree.nwk"

## Assign an Output Directory
output_path <- "./"

## Assign min number of ASVs to prune
min_ASVs = 5

## Comparison Variables
group = "Sample_Name"
control = "Control"

## Beta Diversity Variables
dist = "unifrac"
method = "NMDS"
pvalue_cutoff = 0.05
micromet = "adonis"
pair = TRUE

## Taxonomic Variables
tax_level <- "Class" 

primary_distance_measure <- "Shannon"

```

## Analysis of `r analysis_name`


**Note: All figures produced as part of this RMarkdown are saved in the specified output directory specified below.**

## User Variables

```{r generate_output_directories, include = FALSE}

# Creating Higher Level Directory
analysis_path = paste(output_path,"/16S_Analysis/",sep = "");analysis_path
dir.create(analysis_path)

# Higher Level OTU Path
otupath = paste(analysis_path,"/OTU/",sep = "");otupath
dir.create(otupath)

# Sub OTU Paths
otupath_otu = paste(analysis_path,"/OTU/otutab",sep = "");otupath_otu
otupath_tax = paste(analysis_path,"/OTU/taxtab",sep = "");otupath_tax
otupath_otunorm = paste(analysis_path,"/OTU/otutab_norm",sep = "");otupath_otunorm
otupath_taxnorm = paste(analysis_path,"/OTU/otutax_norm",sep = "");otupath_taxnorm

dir.create(otupath_otu)
dir.create(otupath_tax)
dir.create(otupath_otunorm)
dir.create(otupath_taxnorm)

# Alpha Diversity
alppath = paste(analysis_path,"/alpha_diversity/",sep = "")
dir.create(alppath)

# Beta Diversity
betapath = paste(analysis_path,"/beta_diversity/",sep = "")
dir.create(betapath)

# Taxonomy 
taxapath = paste(analysis_path,"/taxonomy/",sep = "")
dir.create(taxapath)
```

```{r library_preperation, include = FALSE}
## R Installs
#BiocManager::install('ANCOMBC')
#BiocManager::install('cowplot')
#BiocManager::install('decontam')
#BiocManager::install('dplyr')
#BiocManager::install('fs')
#BiocManager::install('ggClusterNet')
#BiocManager::install('ggnewscale')
#BiocManager::install('ggplot2')
#BiocManager::install('ggpubr')
#BiocManager::install('ggsignif')
#BiocManager::install('ggstar')
#BiocManager::install('ggtext')
#BiocManager::install('ggthemes')
#BiocManager::install('ggtreeExtra')
#BiocManager::install('ggvenn')
#BiocManager::install('gsl')
#BiocManager::install('grid')
#BiocManager::install('indicspecies')
#BiocManager::install('magrittr')
#BiocManager::install('picante')
#BiocManager::install('RColorBrewer')
#BiocManager::install('tidyverse')
#BiocManager::install('vegan')
#BiocManager::install('svglite')
#BiocManager::install('DESeq2')
#BiocManager::install('ggrepel')
#BiocManager::install('scales')

## Github Install
#remotes::install_github("pmartinezarbizu/pairwiseAdonis/pairwiseAdonis")
#remotes::install_github("taowenmicro/EasyStat")
#remotes::install_github("taowenmicro/ggClusterNet")

## Library Loading
library(ANCOMBC)
library(cowplot)
library(decontam)
library(dplyr)
library(EasyStat)
library(fs)
library(ggClusterNet)
library(ggnewscale)
library(ggplot2)
library(ggpubr)
library(ggsignif)
library(ggstar)
library(ggtext)
library(ggthemes)
library(ggtreeExtra)
library(ggvenn)
library(gsl)
library(grid)
library(indicspecies)
library(magrittr)
library(pairwiseAdonis)
library(picante)
library(RColorBrewer)
library(tidyverse)
library(vegan)
library(rstatix)
library(svglite)
library(phyloseq)
library(ggrepel)
library(DESeq2)
library(scales)
library(plotly)
library(knitr)
library(kableExtra)
```

```{r, echo=FALSE}
library(knitr)
library(kableExtra)

make_table <- function(params, values) {

  stopifnot(length(params) == length(values))

  data.frame(
    Parameter = params,
    Value = values,
    stringsAsFactors = FALSE
  ) |>
    kable(
      format = "html",
      escape = FALSE
    ) |>
    kable_styling(
      full_width = FALSE,
      bootstrap_options = "bordered",
      position = "left",
      html_font = "sans-serif"
    ) |>
    row_spec(
      0,
      bold = TRUE,
      align = "left",
      background = "#2F4052",
      color = "white"
    ) |>
    row_spec(
      1:nrow(data.frame(params)),
      background = "lightgray"
    ) |>
    column_spec(1, width = "75%") |>
    column_spec(2, width = "50%", extra_css = "word-break: break-all;")
}

make_table(
  params = c(
    "Metadata Path",
    "ASV Feature Table Path",
    "Taxonomy Table Path",
    "Phylogeny Path",
    "Output Path",
    "Minimum Sum of ASVs",
    "Comparative Group Variable",
    "Distance Metric",
    "Plotting Method",
    "P-Value Cutoff",
    "Pairwise Analysis",
    "Taxonomic Level"
  ),
  values = c(
    metadata_path,
    otu_table_path,
    taxonomy_table_path,
    tree_path,
    output_path,
    min_ASVs,
    group,
    dist,
    method,
    pvalue_cutoff,
    pair,
    tax_level
  )
)
```

## Generating a Phyloseq Object

```{r generating_phyloseq_object, include = FALSE}

## Generate a Phyloseq Object
# Feature Table #########################################################################
FEATURE_TABLE_CLEAN <- read.table(otu_table_path,
                                  header = TRUE, 
                                  row.names = 1, 
                                  sep = "\t")

FEATURE_TABLE_CLEAN <- as.matrix(FEATURE_TABLE_CLEAN)
FEATURE_TABLE_CLEAN <- otu_table(FEATURE_TABLE_CLEAN, taxa_are_rows = TRUE)

# Taxonomy Table #########################################################################
TAXONOMY_TABLE <- read.table(taxonomy_table_path,
                             sep = "\t", 
                             header = TRUE, 
                             row.names = 1)

TAXONOMY_SPLIT <- strsplit(as.character(TAXONOMY_TABLE$Taxon), ";")

TAXONOMY_SPLIT_FORMATTED <- lapply(TAXONOMY_SPLIT, function(x) {
  length(x) <- 7 
  return(x)
})

TAXONOMY_MATRIX <- as.matrix(do.call(rbind, TAXONOMY_SPLIT_FORMATTED))

# Remove prefixes like "d__", "p__" etc.
TAXONOMY_MATRIX <- sub("^[a-z]__*", "", TAXONOMY_MATRIX)  # removes d__, p__, etc.
TAXONOMY_MATRIX[TAXONOMY_MATRIX == ""] <- NA # now we fill in the remaining empty gaps with NAs so R doesn't get confused

# Set column names (use "Domain" or "Kingdom" as you prefer)
colnames(TAXONOMY_MATRIX) <- c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species")
# Add back the ASV IDs which were lost from our original data frame
rownames(TAXONOMY_MATRIX) <- rownames(TAXONOMY_TABLE)

# Convert to phyloseq tax_table object
TAXONOMY_TABLE <- phyloseq::tax_table(TAXONOMY_MATRIX)

# Metadata ###############################################################################
METADATA_TABLE <- read.csv(metadata_path, 
                            row.names = 1, 
                            na.strings="", 
                            stringsAsFactors = TRUE)


# Match Metadata Sample IDs to those in the feature table
METADATA_TABLE <- METADATA_TABLE[rownames(METADATA_TABLE) %in% colnames(FEATURE_TABLE_CLEAN), ]
METADATA_TABLE$Sample.name <- rownames(METADATA_TABLE)
METADATA_TABLE[[group]] <- as.factor(METADATA_TABLE[[group]])

METADATA_TABLE <- sample_data(METADATA_TABLE)

# Tree ###################################################################################

ROOTED_TREE <- read_tree(tree_path)

# Generate the PhyloSeq Object ###########################################################
PHYLOSEQ_OBJ <- phyloseq(FEATURE_TABLE_CLEAN, TAXONOMY_TABLE, METADATA_TABLE, ROOTED_TREE)

```

##### Summary of Metadata
```{r summary_metadata}
head(sample_data(PHYLOSEQ_OBJ))
```


#### Alpha Rarefraction Curve

```{r generating_alpha_rarefraction, include = FALSE}
# Calculating Rarefraction Curve

set.seed(2006576) # reproducibility

# Use original clean phyloseq object
dat <- PHYLOSEQ_OBJ

# Define depths
lib_sizes <- sample_sums(dat)
min_lib   <- min(lib_sizes)
max_lib   <- max(lib_sizes)

# Start with your candidate depths
depths <- seq(min_lib, max_lib, by = 1000)
sample_var <- group

# Initialize list
richness_list <- list()

for (d in depths) {
  set.seed(2006576)
  rare_dat <- rarefy_even_depth(dat, sample.size = d, rngseed = 2006576, verbose = FALSE, replace = FALSE)

  richness <- estimate_richness(rare_dat, measures = "Observed")
  richness$SampleID <- rownames(richness)
  richness$Depth <- d
  richness_list[[as.character(d)]] <- richness
}

# Combine results
rare_df <- bind_rows(richness_list)

# Add metadata
metadata <- data.frame(sample_data(dat))
metadata$SampleID <- rownames(metadata)
rare_df <- left_join(rare_df, metadata, by = "SampleID")

# Summarize mean & SE per group
summary_df <- rare_df %>%
  group_by(Depth, !!sym(sample_var)) %>%
  mutate(
    Richness_mean = mean(Observed, na.rm = TRUE),
    Richness_se   = sd(Observed, na.rm = TRUE) / sqrt(n())
  ) %>%
  ungroup()
```

```{r plotting_alpha_rarefraction, include = FALSE}

## Plotting the Rarefraction Curve
ORG_PHYLO_RARE <- ggplot(summary_df, aes(x = Depth, y = Richness_mean, color = SampleID)) +
  geom_line(linewidth = 1) +
  scale_color_discrete(labels = summary_df$SampleID) +
  scale_fill_discrete(labels = summary_df$SampleID) +
  scale_x_continuous(breaks = seq(0, 100000, by = 5000))+
  labs(title = "",
       x = "Sequencing Depth (n)",
       y = "Mean Observed ASVs (n)",
       color = "Group",
       fill = "Group") +
  theme_classic()+
  theme(axis.title.x=element_text(size=14), axis.title.y=element_text(size=14, face = "bold"))+
  theme(axis.text.x=element_text(size=13, angle = 35, hjust = 1, vjust = 1), axis.text.y=element_text(size=13))+
  theme(legend.title=element_text(size=14, face = "bold"), legend.text=element_text(size=13))+
  theme(legend.position="right")

FileName <- paste(alppath,"Refraction_Curve", ".svg", sep = "")
ggsave(FileName, ORG_PHYLO_RARE, width = 10, height = 10, limitsize = FALSE)
```

An alpha refraction curve is generated on the unfiltered dataset. The minimum library size was `r min_lib` and the maximum library size was `r max_lib`.

```{r}
ggplotly(ORG_PHYLO_RARE)
```

## Removing Contamination

```{r decontam_remove_contaminants, include = FALSE}
# Remove incredibly low taxa
PHYLOSEQ_OBJ <- filter_taxa(PHYLOSEQ_OBJ, function(x) sum(x) >= min_ASVs, prune = TRUE)

# Add sequencing depth to sample metadata
sample_data(PHYLOSEQ_OBJ)$SequencingDepth <- sample_sums(PHYLOSEQ_OBJ)

yes_vals <- c("Y","y","Yes","yes","negative","neg","true")

# Identify contaminants with the "combined" method
contam_combined <- isContaminant(
  PHYLOSEQ_OBJ,
  method = "combined",
  conc = sample_data(PHYLOSEQ_OBJ)$SequencingDepth,
  neg <- sample_data(PHYLOSEQ_OBJ)[[control]] %in% yes_vals
)

# How many contaminants?
sum(contam_combined$contaminant)

# Extract contaminant taxa as a character vector (NOT a list)
contaminants <- rownames(contam_combined)[contam_combined$contaminant]

# Double-check overlap with your phyloseq object
length(contaminants)
length(intersect(contaminants, taxa_names(PHYLOSEQ_OBJ)))

# Prune contaminants
PHYLOSEQ_OBJ_CLEAN <- prune_taxa(
  !(taxa_names(PHYLOSEQ_OBJ) %in% contaminants),
  PHYLOSEQ_OBJ
)
```

```{r decontam_output, include = FALSE}
# Prior to decontamination
org_ntaxa <- ntaxa(PHYLOSEQ_OBJ)

# Following decontamination
decon_ntaxa <- ntaxa(PHYLOSEQ_OBJ_CLEAN)

removed_ntaxa <- org_ntaxa - decon_ntaxa
```

The decontam package was used in a combination method taking into account the sequencing depth and the community of the controls. The number of unique taxa/ASVs before decontamination is `r org_ntaxa`. A sum of `r removed_ntaxa` were removed, leaving `r decon_ntaxa` unique taxa/ASVs. 

## Basic Information Tables

From this point, controls are no longer needed for analysis and have been removed. The following tables include basic information (such as abundance/raw count tables are each taxonomic level). These tables can be found here: `r otupath`

```{r}
sample_ctrl <- "Control"
```

```{r}
PHYLOSEQ_OBJ <- PHYLOSEQ_OBJ_CLEAN
sample_data(PHYLOSEQ_OBJ)[[sample_ctrl]] <- as.character(sample_data(PHYLOSEQ_OBJ)[[sample_ctrl]])
PHYLOSEQ_OBJ <- subset_samples(PHYLOSEQ_OBJ, get(sample_ctrl) != "Y")
table(sample_data(PHYLOSEQ_OBJ)[[sample_ctrl]])
```

```{r basic_info_table, include=FALSE}
# Basic Information

## Output OTU Table
otu = as.data.frame(t(ggClusterNet::vegan_otu(PHYLOSEQ_OBJ)))
FileName <- paste(otupath,"/otutab.csv", sep = "")
write.csv(otu,FileName,sep = ",")

## Output Taxa Table
tax = as.data.frame((ggClusterNet::vegan_tax(PHYLOSEQ_OBJ)))
FileName <- paste(otupath,"/tax.csv", sep = "")
write.csv(otu,FileName,sep = "")

## Output Relative Abundance Table
PHYLOSEQ_OBJ_rela  = phyloseq::transform_sample_counts(PHYLOSEQ_OBJ, function(x) x / sum(x) );PHYLOSEQ_OBJ_rela 
otu_norm = as.data.frame(t(ggClusterNet::vegan_otu(PHYLOSEQ_OBJ_rela)))
FileName <- paste(otupath,"/otutab_norm.csv", sep = "")
write.csv(otu_norm,FileName,sep = "")

## Output Relative Abundance Taxa Table
otutax <- cbind(as.data.frame(t(ggClusterNet::vegan_otu(PHYLOSEQ_OBJ_rela))),as.data.frame((ggClusterNet::vegan_tax(PHYLOSEQ_OBJ_rela))))
FileName <- paste(otupath,"/otutax_norm.csv", sep = "")
write.csv(otutax,FileName,sep = "")

## Generate Relative Abundance and Counts OTUs and Taxa for each taxonomic level
for (i in 2: length(phyloseq::rank_names(PHYLOSEQ_OBJ))) {
  psi  <- ggClusterNet::tax_glom_wt(ps = PHYLOSEQ_OBJ,ranks = phyloseq::rank_names(PHYLOSEQ_OBJ)[i])
  #--raw otu tab
  otu = as.data.frame(t(ggClusterNet::vegan_otu(psi)))
  FileName <- paste(otupath_otu,'/',phyloseq::rank_names(PHYLOSEQ_OBJ)[i],".csv", sep = "")
  write.csv(otu,FileName,sep = "")
  # tax table
  tax = as.data.frame((ggClusterNet::vegan_tax(PHYLOSEQ_OBJ)))
  FileName <- paste(otupath_tax,'/',phyloseq::rank_names(PHYLOSEQ_OBJ)[i],".csv", sep = "")
  write.csv(otu,FileName,sep = "")
  
  psi_rela  = phyloseq::transform_sample_counts(psi, function(x) x / sum(x) );psi_rela 
  #--norm otu tab
  otu_norm = as.data.frame(t(ggClusterNet::vegan_otu(psi_rela)))
  FileName <- paste(otupath_otunorm,'/',phyloseq::rank_names(psi)[i],".csv", sep = "")
  write.csv(otu_norm,FileName,sep = "")
  
  otutax <- cbind(as.data.frame(t(ggClusterNet::vegan_otu(psi_rela))),as.data.frame((ggClusterNet::vegan_tax(psi_rela))))
  FileName <- paste(otupath_taxnorm,'/',phyloseq::rank_names(PHYLOSEQ_OBJ)[i],".csv", sep = "")
  write.csv(otutax,FileName,sep = "")
}

```

#### Output OTU Table
```{r summarise_basic_tables0, echo=FALSE}
head(otu)
```

#### Output Taxa Table
```{r summarise_basic_tables1, echo=FALSE}
head(tax)
```

#### Output Relative Abundance Table
```{r summarise_basic_tables2, echo=FALSE}
head(otutax)
```

#### Output Relative Abundance Taxa Table
```{r summarise_basic_tables3, echo=FALSE}
head(otu)
```

## Alpha Diversity

```{r generate_alpha_diversity}
## Perform Diversity Measures

set.seed(2006576)

axis_order =  phyloseq::sample_data(PHYLOSEQ_OBJ)[[group]] %>%unique();axis_order
  ps11 <-  PHYLOSEQ_OBJ
  mapping = phyloseq::sample_data(ps11)
  ps11 = phyloseq::filter_taxa(ps11, function(x) sum(x ) >0 , TRUE); ps11
  colnames(mapping) = gsub(group,"Comparative_Group", colnames(mapping))
  mapping$Group = mapping$Comparative_Group
  mapping$Group = as.factor(mapping$Group)

  # mapping$Group
  count = as.data.frame(t(ggClusterNet::vegan_otu(ps11)))
  alpha=vegan::diversity(count, "shannon")
  x = t(count)

  Shannon = vegan::diversity(x)
  Inv_Simpson <- vegan::diversity(x, index = "invsimpson")
  S <- vegan::specnumber(x);S
  S2 = rowSums(x>0)

Pielou_evenness <- Shannon/log(S)
  Simpson_evenness <- Inv_Simpson/S
  est <- vegan::estimateR(x)
  est <- vegan::estimateR(x)
  Richness <- est[1, ]
  Chao1 <- est[2, ]
  ACE <- est[4, ]
  report = cbind(Shannon, Inv_Simpson, Pielou_evenness, Simpson_evenness,
                 Richness, Chao1,ACE)
  head(report)

alp = merge(mapping,report , by="row.names",all=F)
index = c("Shannon","Inv_Simpson","Pielou_evenness","Simpson_evenness" ,"Richness" ,"Chao1","ACE" )
index = alp
sel = c(match("Shannon",colnames(index)),
        match("Inv_Simpson",colnames(index)),
        match("Pielou_evenness",colnames(index)),
        match("Simpson_evenness",colnames(index)),
        match("Richness",colnames(index)),
        match("Chao1",colnames(index)),
        match("ACE",colnames(index)))

data = cbind(data.frame(ID = 1:length(index$Group),group = index$Group),index[sel])
head(index)

result = EasyStat::MuiaovMcomper2(data = data,num = c(3:9))

FileName <- paste(alppath,"/alpha_diversity_different_label.csv", sep = "")
write.csv(result,FileName,sep = "")
FileName <- paste(alppath,"/alpha_diversity_index.csv", sep = "")
write.csv(index,FileName,sep = "")
```
```{r generating_alpha_plot, include = FALSE}
 ## Generate a Alpha Box Plot

 result1 = EasyStat::FacetMuiPlotresultBox(data = data,num = c(3:9),
                                 result = result,
                                 sig_show ="line",ncol = 3 )
```

#### Alpha Diversity Plot (All Distances)

Alpha Diversity measures are a comparison of in-group diversity (e.g., looking at the differences within certain sample groups).  
The different diversity measures are:

* ACE (Abundance-based Coverage Estimator)
* Chao1 (non-parametric estimator for total species richness)
* Inverse Simpson (1 / Simpson’s Index, measures richness and evenness)
* Simpson Evenness
* Pielou Evenness (evenness of species abundances)
* Richness (absolute number of species)
* Shannon (species abundance + evenness)

```{r stat_plot, fig.width=10, fig.height=10,echo=FALSE}


df <- result1[[2]] %>%
  mutate(dd = suppressWarnings(as.numeric(dd)))


df <- df %>%
  mutate(group = factor(group, levels = sort(unique(group), ignore.case = TRUE)))


# Compute pairwise Wilcoxon tests per facet (name) across groups
pairwise_stats <- df %>%
  group_by(name) %>%
  filter(n_distinct(group[!is.na(dd)]) >= 2) %>%      # keep only facets with >=2 groups
  pairwise_wilcox_test(dd ~ group, p.adjust.method = "fdr") %>%
  ungroup()

# Add bracket positions per facet so labels don’t overlap
pairwise_stats <- pairwise_stats %>%
  group_by(name) %>%
  add_x_position(x = "group") %>%
  add_y_position(data = df, formula = dd ~ group, step.increase = 0.12) %>%
  ungroup()

p1_1 <- ggplot(df, aes(x = group, y = dd)) +
  geom_boxplot(aes(fill = group), alpha = 1, outlier.shape = NA) +
  geom_point(aes(color = group), size = 3, alpha = 0.5) +
  labs(x = "", y = "") +
  facet_wrap(~ name, scales = "free_y", ncol = 3) +
  scale_x_discrete(drop = FALSE) +
  scale_y_continuous(expand = expansion(mult = c(0.1, 0.1))) +
  guides(color = guide_legend(title = NULL),
         shape = guide_legend(title = NULL),
         fill  = guide_legend(title = NULL)) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

# Annotate with precomputed stats (this only draws; it does not recompute)
p1_1 <- p1_1 +
  stat_pvalue_manual(
    pairwise_stats,
    label = "p.adj.signif",     # or "p.adj" for numeric p-values
    hide.ns = TRUE,
    tip.length = 0.01
  )

# Save
FileName <- paste(alppath, "Alpha_box", ".svg", sep = "")
ggsave(FileName, p1_1, width = 15, height = 15, limitsize = FALSE)
```
##### This PLOT
```{r bxcbxcplotting_diversity_to_RMD, fig.width=10, fig.height=10,echo=FALSE}
 p1_1
```

**NB:** if no error bars appear on the plots, no significant comparison was identified. 

```{r}

df_single <- df %>% 
  filter(name == primary_distance_measure)

pairwise_single <- pairwise_stats %>% 
  filter(name == primary_distance_measure)


p_single <- ggplot(df_single, aes(x = group, y = dd)) +
  geom_boxplot(aes(fill = group), alpha = 1, outlier.shape = NA) +
  geom_point(aes(color = group), size = 3, alpha = 0.5) +
  labs(x = "", y = primary_distance_measure) +
  scale_x_discrete(drop = FALSE) +
  scale_y_continuous(expand = expansion(mult = c(0.1, 0.1))) +
  guides(color = guide_legend(title = NULL),
         shape = guide_legend(title = NULL),
         fill  = guide_legend(title = NULL)) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))


p_single <- p_single +
  stat_pvalue_manual(
    pairwise_single,
    label = "p.adj.signif",
    hide.ns = TRUE,
    tip.length = 0.01
  )

```

```{r bcbxcplotting_diversity_to_RMD, fig.width=10, fig.height=10,echo=FALSE}
p_single
```

**NB:** if no error bars appear on the plots, no significant comparison was identified. 

#### Statistical Analysis (Kruskal-Wallis)

```{r}
kw <- kruskal.test(dd ~ group, data = df)
kw
```

```{r post_hoc_test}
if (kw$p.value > 0.05) {
  message("No significant difference identified (KW p = ",
          signif(kw$p.value, 3),
          "). Will not perform a post-hoc test.")
  dunn_res <- NULL
} else {
  dunn_res <- dunn_test(data = df, dd ~ group, p.adjust.method = "fdr")
  print(dunn_res)
}
```


#### Statistical Analysis
```{r ANOVA_on_Alpha}
 ## Perform and Extract Comparison Information

 # ANOVA for each metric

 anova_Shannon <- anova_test(data = index, Shannon ~ Comparative_Group) %>%
   get_anova_table() %>%
   as_tibble() %>%                # Convert to tibble
   mutate(metric = "Shannon")

 anova_InvSimpson <- anova_test(data = index, Inv_Simpson ~ Comparative_Group) %>%
   get_anova_table() %>%
   as_tibble() %>%
   mutate(metric = "Inv_Simpson")

 anova_Pielou <- anova_test(data = index, Pielou_evenness ~ Comparative_Group) %>%
   get_anova_table() %>%
   as_tibble() %>%
   mutate(metric = "Pielou_evenness")

 anova_Evenness <- anova_test(data = index, Simpson_evenness ~ Comparative_Group) %>%
   get_anova_table() %>%
   as_tibble() %>%
   mutate(metric = "Simpson_evenness")

 anova_Richness <- anova_test(data = index, Richness ~ Comparative_Group) %>%
   get_anova_table() %>%
   as_tibble() %>%
   mutate(metric = "Richness")

 anova_Chao1 <- anova_test(data = index, Chao1 ~ Comparative_Group) %>%
   get_anova_table() %>%
   as_tibble() %>%
   mutate(metric = "Chao1")

 anova_ACE <- anova_test(data = index, ACE ~ Comparative_Group) %>%
   get_anova_table() %>%
   as_tibble() %>%
   mutate(metric = "ACE")

 # Combine ANOVA results
 anova_all <- bind_rows(anova_Shannon, anova_InvSimpson, anova_Pielou, anova_Evenness,
                        anova_Richness, anova_Chao1, anova_ACE)

 posthoc_all <- bind_rows(
   tukey_hsd(index, Shannon ~ Comparative_Group) %>% mutate(metric = "Shannon"),
   tukey_hsd(index, Inv_Simpson ~ Comparative_Group) %>% mutate(metric = "Inv_Simpson"),
   tukey_hsd(index, Pielou_evenness ~ Comparative_Group) %>% mutate(metric = "Pielou_evenness"),
   tukey_hsd(index, Simpson_evenness ~ Comparative_Group) %>% mutate(metric = "Simpson_evenness"),
   tukey_hsd(index, Richness ~ Comparative_Group) %>% mutate(metric = "Richness"),
   tukey_hsd(index, Chao1 ~ Comparative_Group) %>% mutate(metric = "Chao1"),
   tukey_hsd(index, ACE ~ Comparative_Group) %>% mutate(metric = "ACE")
 )


 FileName <- paste(alppath,"/alpha_diversity_anova_table.csv", sep = "")
 write.csv(anova_all,FileName,sep = "")
 FileName <- paste(alppath,"/alpha_diversity_tukey_post_hoc.csv", sep = "")
 write.csv(posthoc_all,FileName,sep = "")

```

#### ANOVA Table
```{r, echo=FALSE}
 anova_all
```

#### Post-Hoc Test
```{r ,echo=FALSE}
 posthoc_all
```

## Beta Diversity

A beta diversity ordination plot is generated using the `r method` method and the `r dist` distance. 

```{r ordination_plot, include = FALSE}

ord_df <- ordinate(PHYLOSEQ_OBJ, 
                  method = method, 
                  distance = dist,
                  weighted = TRUE)

stress_val <- ord_df$stress

ord_plot <- plot_ordination(PHYLOSEQ_OBJ, 
                            ord_df,
                            color = group,
                            axes = c(1,2)) +
  stat_ellipse(type = "t", 
               linetype = 2, 
               level = 0.95) +
  ggtitle(sprintf("Method: %s & Distance: %s", method, dist)) +
  geom_point(size = 3) + 
  theme_bw()

FileName <- paste(betapath,"/", method, "_ordination.svg", sep = "")
ggsave(FileName, ord_plot, width = 20, height = 10, limitsize = FALSE)
```

The following plot is coloured by **`r group`**. The stress of the `r method` was: `r stress_val`. 

```{r display_ordination_plot, echo=FALSE, fig.align='center'}
ggplotly(ord_plot)
```

The following plot is coloured by **Sequencing Depth** to assure no confounding grouping. 

```{r generating_beta_diversity_measure}
ord_plot <- plot_ordination(PHYLOSEQ_OBJ, 
                            ord_df,
                            color = "SequencingDepth") +
  ggtitle(sprintf("Method: %s & Distance: %s", method, dist)) +
  geom_point(size = 3) + 
  theme_bw()

FileName <- paste(betapath,"/", method, "SeqDepth_ordination.svg", sep = "")
ggsave(FileName, ord_plot, width = 20, height = 10, limitsize = FALSE)
```

```{r plot_ordination, fig.align='center',echo=FALSE}
ggplotly(ord_plot)
```

#### Interactive 3D Plot

## Taxonomic Analysis

```{r taxanomic_dot, include=FALSE,echo=FALSE}
#  # Collapse to chosen level
#  PHYLOSEQ_OBJ_TAXA <- tax_glom(PHYLOSEQ_OBJ, taxrank = tax_level)
# 
#  # Relative abundance
#  PHYLOSEQ_OBJ_RELATIVE_ABUND <- transform_sample_counts(PHYLOSEQ_OBJ_TAXA, function(x) x / sum(x))
# 
#   #Melt
#  PHYLOSEQ_OBJ_PLOT <- psmelt(PHYLOSEQ_OBJ_RELATIVE_ABUND)
# 
#  # ---- NEW PART: Identify top 20 taxa ----
# 
#  top_taxa <- PHYLOSEQ_OBJ_PLOT %>%
#    as_tibble() %>%  # ensure a tibble, not an S4/IRanges object
#    group_by(.data[[tax_level]]) %>%
#    summarise(mean_abundance = mean(Abundance, na.rm = TRUE), .groups = "drop") %>%
#    arrange(desc(mean_abundance)) %>%
#    dplyr::slice(1:30) %>%
#    dplyr::pull(.data[[tax_level]]) %>%
#    unique()
# 
# 
#  # ---- Recode everything else as "Other" ----
#  PHYLOSEQ_OBJ_PLOT <- PHYLOSEQ_OBJ_PLOT %>%
#    mutate(!!sym(tax_level) := if_else(!!sym(tax_level) %in% top_taxa,
#                                       !!sym(tax_level),
#                                       "Other"))
# 
# 
# 
# PHYLOSEQ_OBJ_PLOT <- PHYLOSEQ_OBJ_PLOT %>%
#    group_by(!!sym(group), !!sym(tax_level)) %>%
#    summarise(Abundance = sum(Abundance), .groups = "drop")
# 
# 
# 
#  taxa_order <- c(top_taxa, "Other")
#  PHYLOSEQ_OBJ_PLOT[[tax_level]] <- factor(PHYLOSEQ_OBJ_PLOT[[tax_level]], levels = taxa_order)
# 
#  # Filter out abundances below 0.1
#  PHYLOSEQ_OBJ_PLOT <- PHYLOSEQ_OBJ_PLOT %>%
#    filter(Abundance >= 0.01) # Do not show the rare samples
# 
#  ## Wrap the Facet Labels
#  facet_labels_wrapped <- setNames(
#    stringr::str_wrap(levels(PHYLOSEQ_OBJ_PLOT[[group]]), width = 10),
#    levels(PHYLOSEQ_OBJ_PLOT[[group]])
#  )
# 
#  p4 <- ggplot(PHYLOSEQ_OBJ_PLOT,
#         aes(x = SampleID,
#             y = !!sym(tax_level),
#           size = Abundance,
#           color = !!sym(tax_level))) +
#  geom_point(alpha = 0.7) +
#  theme_bw() +
#  facet_wrap(~ type, scales = 'free_x', nrow = 1) +
#  theme(axis.text.x = element_text(angle = 90, vjust = 0.5)) +
#  ylab(tax_level) +
#  xlab("") +
#  scale_size_continuous(range = c(1, 20)) +
#    guides(color = guide_legend(nrow = 20, byrow = FALSE, override.aes = list(size = 6))) +
#    theme(strip.text = element_text(size = 12, face = "bold")) +
#    theme(axis.title.x=element_text(size=12), axis.title.y=element_text(size=12)) +
#    theme(axis.text.x=element_text(size=12), axis.text.y=element_text(size=12)) +
#    theme(legend.title=element_text(size=12, face = "bold"), legend.text=element_text(size=12)) +
#    theme(legend.position="right")
# 
# 
#  FileName <- paste(alppath,"Dot_Taxonomy.svg", sep = "")
#  ggsave(FileName, p4, width = 30, height = 10, limitsize = FALSE)
```

```{r plotting_taxonomy, fig.width=10, fig.height=20, echo=FALSE}
#ggplotly(p4, width = 1500, height = 1000)
```

## Differential Abundance


## Functional Prediction


